### ISSUE TITLE
Unnecessary declarations of builtin functions as external C libraries in Math

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#931]
**Reported by pharman on 14 Dec 2012 13:19 UTC**
The annotations in Modelica/Math/package.mo on sin, cos, tan, asin, acos, atan, atan2, sinh, cosh, tanh, exp, log & log10 aren't required because they are all declared as "builtin" not "C" so a Library annotation is not relevant.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/931

**Comment by modelica-trac-importer** [comment]
**Comment by pharman on 14 Dec 2012 16:15 UTC**
This part is trivial, the formatting as described and fixed in #929 was the higher priority issue.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 2 Jan 2013 21:42 UTC**
Built-in functions defined in MSL are currently declared as:

```
function sin "Sine"
  extends Modelica.Math.Icons.AxisLeft;
  input Modelica.SIunits.Angle u;
  output Real y;
  external"builtin" y = sin(u) annotation (Library="ModelicaExternalC");
end sin; 
```
This is correct Modelica 3 and allows two ways that a tool can handle this information:

 1. Ignore the external prototype and the Library annotation and handle the function specially in the tool.
 1. Do not handle the function specially with the exception to interpret "builtin" as "C" and library "ModelicaExternalC" as the ANSI C standard library.

If all tool vendors utilize approach (1), it seems best to remove the unnecessary prototype and the Library annotation. If there are tool vendors that utilize approach (2), it seems best to just keep the current definition.

To summarize: The tool vendors should give their opinions what to do.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 3 Jan 2013 06:18 UTC**
If (2) is used, Library should of course be "m" (to link to the math library, libm, on Linux), or removed. The ModelicaExternalC sources do not contain any function `sin`.

I would not say the sources are correct Modelica 3 because actually "builtin" is implementation defined... And the only example in the text uses `external "builtin";`. Besides, if a tool should ignore the prototype and annotation, why keep them there?

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 3 Jan 2013 10:13 UTC**
The library annotation seems odd. In fact, I would say "builtin" functions should belong into a core library (being part of the specification) and not the standard library. A Modelica tool should be able to work without the MSL IMO.

The benefit of having the function signatures as externals is twofold:

 1. There is a machine readable, canonical form of builtin functions. It is highly unlikely a tool misses one of them or implements them wrong.
 2. It saves implementation time. For type checking, the whole signature needs to exist in a tool anyway. So why force everybody to re-implement them, when there is a way to write them in pure Modelica?

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 3 Jan 2013 10:55 UTC**
Replying to [comment:4 choeger]:
> A Modelica tool should be able to work without the MSL IMO.

It does. The MSL function Modelica.Math.sin depends on the builtin function `sin`, which is in the specification.


Forcing a prototype for a builtin function seems odd, since our builtin sin function may be called as `fast_sin(in,out)` instead of `out = sin(in)` (for the record: it doesn't). I thought this was why the functions were changed from C to builtin in the first place.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 3 Jan 2013 14:35 UTC**
Replying to [comment:5 sjoelund.se]:
> Replying to [comment:4 choeger]:
> > A Modelica tool should be able to work without the MSL IMO.
> 
> It does. The MSL function Modelica.Math.sin depends on the builtin function `sin`, which is in the specification.
> 
> 
> Forcing a prototype for a builtin function seems odd, since our builtin sin function may be called as `fast_sin(in,out)` instead of `out = sin(in)` (for the record: it doesn't). I thought this was why the functions were changed from C to builtin in the first place.

The point is, that the spec may contain bugs and/or holes when it comes to the specific signature of builtin functions. This is highly unlikely in case of formal prototype definitions. This might seem unlikely in case of `sin` but it also does not hurt to formally specify simple stuff. IMO everything that can be specified using Modelica should be (mainly because it saves a lot of effort). 

And btw: omc already contains such a "core": Compiler/FrontEnd/ModelicaBuiltin.mo 

ps: I do not really understand what you mean with that `fast_sin`.

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 26 Feb 2013 21:47 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 20 May 2013 08:30 UTC**
Replying to [comment:3 sjoelund.se]:
> If (2) is used, Library should of course be "m" (to link to the math library, libm, on Linux), or removed. The ModelicaExternalC sources do not contain any function `sin`.

Removed in that case. 

The idea of separate linking with "libm" is just old (from the times when avoiding linking with a MB file was considered a significant saving), and not supported by all compilers (and some compilers have an empty libm.a just to avoid problems.

We are allowed to write `x^y` in Modelica without specifying anything, even though this will also likely require linking with math-functions.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 20 May 2013 09:03 UTC**
Fixed in ebdf7ba332cf4f9d1b4a3c6105433abea2df6297 and in r6406:

Removed the wrong ModelicaExternalC library annotation from builtin functions and added in the Modelica 3.2r2 specification a clarification:

The "builtin" specification is only used for functions that are defined to be built-in in Modelica. The external-function call mechanism for "builtin" functions is implementation-defined. <ins>_Typically, for functions from the standard C-library, the prototype of the function is provided but no library annotation. Currently, there are no other builtin functions defined in Modelica._</ins>] 
[Example:

```
package Modelica 
  package Math
    function sin
      input Real x;
      output Real y;
      external "builtin" y=sin(x);
    end sin;
  end Math;
end Modelica;
```
